### ISSUE TITLE
There should not be more than one version of branch MAIN/SNOMEDCT-NO with no end date.

### ISSUE THREAD

**Issue by Zawaken** [IHTSDO/snowstorm#320]
Hi,
Environment:
2 servers with snowstorm running, load balanced.
2 elasticsearch instances in a cluster.
Daily-build enabled on the MAIN/SNOMEDCT-NO branch (these only get downloaded/loaded on one of the servers.)

Stacktrace:
```
Sep 17 13:31:43 java[17680]: java.lang.IllegalStateException: There should not be more than one version of branch MAIN/SNOMEDCT-NO with no end date.
Sep 17 13:31:43  java[17680]: at io.kaicode.elasticvc.api.BranchService.illegalState(BranchService.java:596)
Sep 17 13:31:43 java[17680]: at io.kaicode.elasticvc.api.BranchService.findLatest(BranchService.java:128)
Sep 17 13:31:43 java[17680]: at org.snomed.snowstorm.core.rf2.rf2import.ImportService.importArchive(ImportService.java:99)
Sep 17 13:31:43 java[17680]: at org.snomed.snowstorm.core.rf2.rf2import.ImportService$$FastClassBySpringCGLIB$$3479d1da.invoke(<generated>)
Sep 17 13:31:43 java[17680]: at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
Sep 17 13:31:43 java[17680]: at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:687)
Sep 17 13:31:43 java[17680]: at org.snomed.snowstorm.core.rf2.rf2import.ImportService$$EnhancerBySpringCGLIB$$aab2f7df.importArchive(<generated>)
Sep 17 13:31:43 java[17680]: at org.snomed.snowstorm.dailybuild.DailyBuildService.dailyBuildDeltaImport(DailyBuildService.java:103)
Sep 17 13:31:43 java[17680]: at org.snomed.snowstorm.dailybuild.DailyBuildService.performScheduledImport(DailyBuildService.java:85)
Sep 17 13:31:43 java[17680]: at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Sep 17 13:31:43 java[17680]: at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
Sep 17 13:31:43 java[17680]: at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Sep 17 13:31:43 java[17680]: at java.base/java.lang.reflect.Method.invoke(Method.java:566)
Sep 17 13:31:43 java[17680]: at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
Sep 17 13:31:43 java[17680]: at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
Sep 17 13:31:43 java[17680]: at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:774)
Sep 17 13:31:43 java[17680]: at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
Sep 17 13:31:43 java[17680]: at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:749)
Sep 17 13:31:43 java[17680]: at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:691)
Sep 17 13:31:43 java[17680]: at org.snomed.snowstorm.dailybuild.DailyBuildService$$EnhancerBySpringCGLIB$$a6201fa2.performScheduledImport(<generated>)
Sep 17 13:31:43 java[17680]: at org.snomed.snowstorm.dailybuild.ScheduledDailyBuildImportService.scheduledDailyBuildDeltaImport(ScheduledDailyBuildImportService.java:52)
Sep 17 13:31:43 java[17680]: at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Sep 17 13:31:43 java[17680]: at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
Sep 17 13:31:43 java[17680]: at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Sep 17 13:31:43 java[17680]: at java.base/java.lang.reflect.Method.invoke(Method.java:566)
Sep 17 13:31:43 java[17680]: at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:84)
Sep 17 13:31:43 java[17680]: at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
Sep 17 13:31:43 java[17680]: at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
Sep 17 13:31:43 java[17680]: at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
Sep 17 13:31:43 java[17680]: at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
Sep 17 13:31:43 java[17680]: at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
Sep 17 13:31:43 java[17680]: at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
Sep 17 13:31:43 java[17680]: at java.base/java.lang.Thread.run(Thread.java:829)
```

We're having an issue where one of our environments are errorring out with the error in the title.
This might have first happened when a rollback was done.
We've since restored the servers from a VM snapshot twice, and it has happened again both times.

From the logs before the stacktrace it might actually look like that is what has happened this time as well:
```
Sep 17 13:31:37 java[17680]: 2021-09-17 13:31:37.541  INFO 17680 --- [   scheduling-1] o.s.s.dailybuild.DailyBuildService       : Commit 1631683165600 does not have a new base (1629896193528) so will be rolled back.
Sep 17 13:31:37 java[17680]: 2021-09-17 13:31:37.541  INFO 17680 --- [   scheduling-1] o.s.s.dailybuild.DailyBuildService       : 1 branch commits found to roll back on MAIN/SNOMEDCT-NO.
Sep 17 13:31:37 java[17680]: 2021-09-17 13:31:37.541  INFO 17680 --- [   scheduling-1] io.kaicode.elasticvc.api.BranchService   : Rolling back commit 1631683165600 on MAIN/SNOMEDCT-NO.
Sep 17 13:31:43 java[17680]: 2021-09-17 13:31:43.439  INFO 17680 --- [   scheduling-1] io.kaicode.elasticvc.api.BranchService   : Completed rollback of commit 1631683165600 on MAIN/SNOMEDCT-NO.
```

I am ***assuming*** this might be a problem that occured because we have set up two snowstorm servers load balanced that access the same cluster.

**Comment by kaicode** [comment]
Hi.
Having two Snowstorm instances connected to the same Elasticsearch cluster should be fine as long as only one of the instances is writing data. The dailybuild import is disabled as default so as long as only one Snowstorm instance has:
```
daily-build.delta-import.enabled=true
```
([this property](https://github.com/IHTSDO/snowstorm/blob/7.3.0/src/main/resources/application.properties#L740)) I would not expect an issue.

The log messages before the stacktrace look normal. The dailybuild process rolls back commits on the code system branch until it finds one with a different base timepoint. A commit with a different base indicates that it could be a code system upgrade commit and those should not be rolled back automatically.

Are the two Elasticsearch instances in the same data center? Although having them in different data centers / regions sounds like good resilience architecture it may introduce too much latency in the store for updates which could cause problems.

There is no easy way to resolve this issue. There is an admin function to hard delete a branch and all it's history which would avoid reimporting the content on other branches (like MAIN). The alternative is to drop your indices and start again.

**Comment by Zawaken** [comment]
Both of the elasticsearches (running on the snowstorm servers), are in the same data center.

I do think both of them are able to write to the cluster, I don't think this should be an issue.
Only one of them has the `daily-build.delta-import.enabled` option set to `true`.
From what I've gathered 'MAIN/SNOMEDCT-NO' (and all children of that branch) are the only ones that are duplicated.

I checked the other server's logs, and I see that it hadn't seen the commit to rollback.
So my assumption was likely wrong.

Let me know if there are any full logs that I should give you or if there is additional info needed.
